[Easee] Initial contribution (#12954)
openhab/openhab-addons: bundles/org.openhab.binding.easee/README.md

# Easee Binding

The Easee binding can be used to retrieve data from the Easee Cloud API and also to control your wallbox via the Cloud API.
This allows you to dynamically adjust the charge current for your car depending on production of your solar plant.

## Supported Things

This binding provides three thing types: 

| Thing/Bridge        | Thing Type          | Description                                                                                   |
|---------------------|---------------------|-----------------------------------------------------------------------------------------------|
| bridge              | site                | cloud connection to a site within an Easee account                                            |
| thing               | charger             | the physical charger which is connected to a circuit within the given site                    |
| thing               | mastercharger       | like the "normal" charger but with additional capability to control the circuit               |

Basically any Easee wallbox that supports the Cloud API should automatically be supported by this binding.

## Discovery

> Discovery is implemented and will discover all circuits and chargers assigned to a given site.

Author: fixed

|-------------------------|----------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| username                | yes      | The username to login at Easee Cloud service. This should be an e-mail address or phone number.                                                                                                                                                                                                                                                          |
| passord                 | yes      | Your password to login at Easee Cloud service.                                                                                                                                                                                                                                                                                                           |

> | siteId                  | yes      | The ID of the site containing the wallbox(es) and circuit(s) that should be integrated into openHAB. The ID of your site can be found out via the sites overview (https://easee.cloud/sites). You just need to click on one of the sites listed there, the id will be part of the URL which is then opened. It will be a number with typically 6 digits. |

Reviewer: ```suggestion
| siteId                  | yes      | The ID of the site containing the wallbox(es) and circuit(s) that should be integrated into openHAB. The ID of your site can be found via the sites overview (https://easee.cloud/sites). You just need to click one of the sites listed there, the id will be part of the URL which is then opened. It will be a number with typically 6 digits. |
```

Author: fixed

> | dataPollingInterval     | no       | interval (seconds) in which live data values are retrieved from the Easee Cloud API. (default = 120)                                                                                                                                                                                                                                                     |

Reviewer: ```suggestion
| dataPollingInterval     | no       | Interval (seconds) in which live data values are retrieved from the Easee Cloud API. (default = 120)                                                                                                                                                                                                                                                     |
```

Author: fixed

### Charger

| Configuration Parameter | Required | Description                                                                                                            |
|-------------------------|----------|------------------------------------------------------------------------------------------------------------------------|
| id                      | yes      | The id of the charger that will be represented by this thing.                                                          |

### Mastercharger

| Configuration Parameter | Required | Description                                                                                                            |
|-------------------------|----------|------------------------------------------------------------------------------------------------------------------------|
| id                      | yes      | The id of the charger that will be represented by this thing.                                                          |
| circuitId               | yes      | The id of the circuit that is controlled by this charger.                                                              |

## Channels

> The binding only supports a subset of the available endpoints that is provided by the Easee Cloud API.

Reviewer: ```suggestion
The binding only supports a subset of the available endpoints provided by the Easee Cloud API.
```

Author: fixed

> The tables below shows all channels that are available and which of them are writable.

Reviewer: ```suggestion
The tables below show all available channels and which of them are writable.
```

Author: fixed

> The setting that start with "dynamic" can be changed frequently, the other are written to flash and thus should not be changed too often as this could result in damage of your flash.

Reviewer: ```suggestion
The settings that start with "dynamic" can be changed frequently, the others are written to flash and thus should not be changed too often as this could result in damage of your flash.
```

Author: fixed

### Charger Channels

> | Channel Type ID                             | Item Type                | Writable | Description                                          | Allowed Values (write access)                                                                                                                                |

Reviewer: Header for first column should be "Channel" or similar. It could also be separated into two columns: Channel and Channel Group. Channel Type ID is not of much relevance to users.

Author: Channel Type ID is what you need when you use textual configuration. Or maybe it is just the wrong naming of the column? However I can separate it into group and channel.

Reviewer: Yes, it's only the naming. I just committed a change to "Channel" in case you wouldn't respond, because I will then fit your PR into next milestone. If you would like to separate into group and channel, please go ahead. I'm leaving in 20 minutes, if it will be quick change, I'll merge your PR by then. :-)

Reviewer: Btw, don't worry about the DCO from my commit, I'll fix it when merging.

| latestSession#sessionEnergy                 | Number:Energy            | no       | latest (already ended) session                       |                                                                                                                                                              |
| latestSession#sessionStart                  | DateTime                 | no       |                                                      |                                                                                                                                                              |
| latestSession#sessionEnd                    | DateTime                 | no       |                                                      |                                                                                                                                                              |

### Master Charger Channels

> The Master Charger is like the "normal" charger but has some extra channel to control the circuit. These additional channels are listed in the table below.

Reviewer: ```suggestion
The Master Charger is like the "normal" charger but has some extra channels to control the circuit. These additional channels are listed in the table below.
```

Author: fixed

> | Channel Type ID                             | Item Type                | Writable | Description                                          | Allowed Values (write access)                                                                                                                                |

Reviewer: Same here.

| settings#offlineMaxCircuitCurrentP3         | Number:ElectricCurrent   | no       |                                                      |                                                                                                                                                              |
| settings#offlineMaxCurrents                 | String                   | yes      | read/write only for all phases.                      | <value phase1>;<value phase2>;<value phase3>  valid values for each phase are 0, 6-32                                                                        |
| settings#enableIdleCurrent                  | Switch                   | yes      |                                                      | true/false                                                                                                                                                   |
| settings#allowOfflineMaxCircuitCurrent      | Switch                   | no       |                                                      |                                                                                                                                                              |

> |

Author: fixed

## Full Example

### Thing                                                                                                                                                                    

- minimum configuration

```
Bridge easee:site:mysite1 [ username="abc@def.net", password="secret", siteId="123456" ]
```

> - manual configuration with two chargers, pollingInterval set to 60 seconds.

Reviewer: ```suggestion
#### Manual configuration with two chargers, pollingInterval set to 60 seconds.
```

Author: fixed

Number:ElectricCurrent  Easee_Circuit_Phase3                  "Phase 3"                                   { channel="easee:mastercharger:mysite1:myCharger1:dynamicCurrent#phase3" }
String                  Easee_Circuit_Dynamic_Phases          "Dynamic Power [MAP(easeePhases.map):%s]"   { channel="easee:mastercharger:mysite1:myCharger1:dynamicCurrent#setDynamicCurrents" }
Switch                  Easee_Charger_Start_Stop              "Start / Stop"                              { channel="easee:mastercharger:mysite1:myCharger1:commands#startStop" }
```

### Sitemap

```
    Switch item=Easee_Circuit_Dynamic_Phases mappings=["0;0;0"="0,00 kW", "6;0;0"="1,44 kW", "7;0;0"="1,68 kW", "8;0;0"="1,92 kW", "9;0;0"="2,16 kW", "10;0;0"="2,40 kW", "16;0;0"="3,72 kW", "16;16;16"="11,1 kW"] icon="energy"
```

### Mapping

easeePhases.map will make the phase setting more readable.

```
0;0;0=0 kW
6;0;0=1,44 kW
7;0;0=1,68 kW
8;0;0=1,92 kW
9;0;0=2,16 kW
10;0;0=2,40 kW
11;0;0=2,64 kW
12;0;0=2,88 kW
13;0;0=3,12 kW
14;0;0=3,36 kW
15;0;0=3,60 kW
16;0;0=3,72 kW
6;6;6=4,32 kW
16;16;16=11,1 kW

> ´´´

Author: fixed

### Mastercharger

| Configuration Parameter | Required | Description                                                                                                            |
|-------------------------|----------|------------------------------------------------------------------------------------------------------------------------|
| id                      | yes      | The id of the charger that will be represented by this thing.                                                          |
| circuitId               | yes      | The id of the circuit that is controlled by this charger.                                                              |

## Channels

The binding only supports a subset of the available endpoints provided by the Easee Cloud API.
The tables below show all available channels and which of them are writable.
The settings that start with "dynamic" can be changed frequently, the others are written to flash and thus should not be changed too often as this could result in damage of your flash.

### Charger Channels

> | Channel Type ID                             | Item Type                | Writable | Description                                          | Allowed Values (write access)                                                                                                                                |

Reviewer: ```suggestion
| Channel                                     | Item Type                | Writable | Description                                          | Allowed Values (write access)                                                                                                                                |
```

| latestSession#sessionEnergy                 | Number:Energy            | no       | latest (already ended) session                       |                                                                                                                                                              |
| latestSession#sessionStart                  | DateTime                 | no       |                                                      |                                                                                                                                                              |
| latestSession#sessionEnd                    | DateTime                 | no       |                                                      |                                                                                                                                                              |

### Master Charger Channels

The Master Charger is like the "normal" charger but has some extra channels to control the circuit. These additional channels are listed in the table below.

> | Channel Type ID                             | Item Type                | Writable | Description                                          | Allowed Values (write access)                                                                                                                                |

Reviewer: ```suggestion
| Channel                                     | Item Type                | Writable | Description                                          | Allowed Values (write access)                                                                                                                                |
```

| thing               | mastercharger       | like the "normal" charger but with additional capability to control the circuit               |

Basically any Easee wallbox that supports the Cloud API should automatically be supported by this binding.

## Discovery

Discovery is implemented and will discover all circuits and chargers assigned to a given site.

## Bridge Configuration

The following configuration parameters are available for the binding/bridge:

- **username** (required)
The username to login at Easee Cloud service.
This should be an e-mail address or phone number.

- **passord** (required)  
Your password to login at Easee Cloud service.

- **siteId** (required)  
The ID of the site containing the wallbox(es) and circuit(s) that should be integrated into openHAB.
The ID of your site can be found out via the sites overview (https://easee.cloud/sites).
You just need to click on one of the sites listed there, the id will be part of the URL which is then opened.
It will be a number with typically 6 digits.

- **dataPollingInterval** (optional)  

> interval (seconds) in which live data values are retrieved from the Easee Cloud API. (default = 120)

Reviewer: I would suggest to format this as a table for better overview/readability, like many other bindings do. See for example https://www.openhab.org/addons/bindings/hdpowerview/#thing-configuration.

## Thing configuration

It is recommended to use auto discovery which does not require further configuration.
If manual configuration is preferred you need to specify configuration as below.

### Charger

- **id** (required)

> The id of the charger that will be represented by this thing.

Reviewer: Same here (see comment above).

Author: fixed

### Mastercharger

- **id** (required)
The id of the charger that will be represented by this thing.

- **circuitId** (required)

> The id of the circuit that is controlled by this charger.

Reviewer: Same here (see comment above).

Author: fixed
